core/src/surplus.ts

```diff
@@ -0,0 +1,88 @@
+import type { SurplusAuction, SurplusAuctionBase } from './types';
+import { getEarliestDate } from './helpers/getEarliestDate';
+import BigNumber from './bignumber';
+import getContract from './contracts';
+import { Contract } from 'ethers';
+import getNetworkDate from './date';
+import { RAD, WAD } from './constants/UNITS';
+
+const getSurplusAuctionLastIndex = async (contract: Contract): Promise<number> => {
+    const auctionsQuantityBinary = await contract.kicks();
+    return new BigNumber(auctionsQuantityBinary._hex).toNumber();
+};
+
+const getAuctionState = async (network: string, earliestEndDate: Date, greatestBid: number) => {
+    const isBidExpired = (await getNetworkDate(network)) > earliestEndDate;
+    const haveBids = greatestBid === 0;
+    if (haveBids) {
+        if (isBidExpired) {
+            return 'ready-for-collection';
+        }
+        return 'have-bids';
+    }
+    if (isBidExpired) {
+        return 'requires-restart';
+    }
+    return 'just-started';
+};
+
+export const fetchSurplusAuctionByIndex = async function (
+    network: string,
+    auctionIndex: number
+): Promise<SurplusAuction> {
+    const contract = await getContract(network, 'MCD_FLAP');
+    const auctionData = await contract.bids(auctionIndex);
+    const isAuctionCollected = new BigNumber(auctionData.end).eq(0);
+    const baseAuctionInfo: SurplusAuctionBase = {
+        network,
+        id: auctionIndex,
+    };
+
+    if (isAuctionCollected) {
+        const auctionLastIndex = await getSurplusAuctionLastIndex(contract);
+        if (auctionLastIndex < auctionIndex) {
+            throw new Error('No active auction exists with this id');
+        }
+        return { ...baseAuctionInfo, state: 'collected' };
+    }
+
+    const auctionEndDate = new Date(auctionData.end * 1000);
+    const bidEndDate = new Date(auctionData.tic * 1000);
+    const earliestEndDate = getEarliestDate(auctionEndDate, bidEndDate);
```

Comment:
Also, `bidEndDate` [is not set](https://github.com/makerdao/dss/blob/fa4f6630afb0624d04a003e920b0d71a00331d98/src/flap.sol#L117-L132) during kick, therefore, `bidEndDate` will be 0. Instead of creating `new Date(0)` which will be `Thu Jan 01 1970 01:00:00`, you should set `bidEndDate` to undefined.

That's why I proposed to have `earliestEndDate` in the first place: to figure out this complex edge case once